### ISSUE TITLE
$customer->getOrigData() return null in customer_save_before event

### ISSUE THREAD

**Issue by hiephm** [magento/magento2#4662]
## Steps to reproduce
1. Create an observer to observe `customer_save_before` event
2. Save customer in backend, check the customer object in observer 

``` php
$customer = $observer->getEvent()->getCustomer();
```
## Expected result

`$customer->getOrigData()` should return data before save, so that we can compare with updated data to perform some logic.
## Actual result

`$customer->getOrigData()` return null because it never loaded.
Currently I have to work-around by load customer again in observer.

**Comment by palamar** [comment]
@hiephm 
Please, provide the used version. If the problem is actual for a specific tag, please, specify it and be sure that the latest update was used.

**Comment by palamar** [comment]
According to [contributor guide](http://devdocs.magento.com/guides/v2.1/contributor-guide/contributing.html), tickets without response for two weeks should be closed.
Please feel free to reopen if it's needed.

**Comment by tools-utrecht-hh** [comment]
@palamar I am having the exact same issue as hiephm, I think it has something to do with the fact that the customer admin controller uses the customer repository to load a customer. Because if I reload the entire customer in the observer via the customer model it contains the OrigData.

We are running on Magento 2.1.0 and are using the same event customer_save_after as hiephm.

**Comment by ksz2013** [comment]
Same issue here as well. Is there a workaround to check if data was saved on a customer ?
As previously mentioned in the observer the OrigData is null... can't use the dataHasChangedFor on the customer object

**Comment by ksz2013** [comment]
@experius-nl have you managed to find a workaround for this? I am having this issue on the frontend as well. Tried a workaround with aroundSave plugin on the resourceModel but same issue.... origData is empty there as wel..

**Comment by hiephm** [comment]
@ksz2013: you need to load customer explicitly in customer_save_before event handler.

**Comment by Weltpixel** [comment]
That's possible, but it is an additional overload, what is the purpose of the origData if it is not containing the proper values?

**Comment by slavvka** [comment]
Could you please clarify do you save a new customer or an existent one?

**Comment by hiephm** [comment]
@slavvka it's existent customer, that why I expected `$customer->getOrigData()` should return original data. This behavior (likely a bug) is simply because customer is never loaded.

**Comment by slavvka** [comment]
The problem must be fixed on the latest mainline. Could you please check that? Soon it will be backported to 2.0 and 2.1

**Comment by veloraven** [comment]
According to [contributor guide](http://devdocs.magento.com/guides/v2.1/contributor-guide/contributing.html), tickets without response for two weeks should be closed.
If this issue still reproducible please feel free to create the new one: format new issue according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines): with steps to reproduce, actual result and expected result and specify Magento version.

**Comment by TheFrankman** [comment]
This still hasn't been resolved in 2.1.1

**Comment by eInyzant** [comment]
This still happen in 2.1.5 :(

**Comment by jaykobi** [comment]
Still happens in 2.1.7

**Comment by legweilo** [comment]
This is fixed on 2.1.2, in your observer:

```
$customerData = $observer->getEvent()->getData('customer');
$newData = $customerData->getData('xxx_data');
$previousData = $customerData->getOrigData('xxx_data');
```

**Comment by milansimek** [comment]
Issue is still present in 2.1.7 and 2.2 rc

**Comment by radub** [comment]
Issue still present in M2.3

**Comment by crazylink** [comment]
Have you planed to fix it ?

**Comment by willstorm** [comment]
Alternatively, if you can use the event after saving the customer. Use the customer_save_after_data_object event. It is defined in **module-customer/Model/ResourceModel/CustomerRepository.php** on line 274.
The parameters it passes are customer_data_object and orig_customer_data_object.

**Comment by matheusgontijo** [comment]
Thanks @willstorm !!!

**Comment by 0-Sony** [comment]
How this issue could have been closed ?? This issue is still present on 2.3.x and I am pretty sure that it's still on 2.4.x 
The solution gave by @willstorm is not a solution.

**Comment by milansimek** [comment]
I agree with @0-Sony. It's not a solution at all.

Loading the customer model in the original event as outlined in the issue description would be a better, although not optimal, workaround. It's a workaround though and not a solution, and does not explain why this issue can be closed.
